### PR TITLE
Fixes #7006: Parameter validation in Rudder should accept rudder variables ( ${ } )

### PR DESCRIPTION
**Pull request by fanf**
https://issues.rudder.io/issues/7006

"oups". 

It was *really* too complicated to add non-rudder variable to our compiler with scala lib parser combintor. Since we now have `fastparse` in the path, I tried to see if it was easier... And it was thanks to its amazing log possibilities (the parsing by itself is a bit tricky). 
The most interesting is how little code changes from scala regex combinator to fastparse. And the perf are just in other league (it does not change much in our case, since node prop interpolation is in second range, but now we are in the 10s of millisec - it may still be interesting with group properties and many more parsing to do). 
So I also switch the last parser we had (the one for our "search anything" bar queries), what allows to get ride of one dependancy and use the (very cool!) same combinator parser everywhere.

### CODE REVIEW

**Comment by fanf** [comment]
PR rebased

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 1]
```diff
@@ -406,64 +405,88 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
   //also parse full blank string, because no need to look for more case for them.
-  val emptyVar    : Parser[List[CharSeq]] = """(?iums)(\s)*""".r ^^ { x => List(CharSeq(x)) }
-
+  def emptyVar[_: P]: P[List[CharSeq]] = P(CharsWhile(_.isSpaceChar, 0).!).map(x => List(CharSeq(x)) )
+  def s[_:P] = P(CharsWhile(_.isWhitespace, 0))
   // plain string must not match our identifier, ${rudder.* and ${node.properties.*}
-  // here we defined a function to build them, with the possibility to
-  val plainString : Parser[CharSeq] = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties)).)+""").r  ^^ { CharSeq(_) }
+  // here we defined a function to build them
+  def plainStringNoDollar[_: P] : P[CharSeq] = P( (!"${" ~ AnyChar).rep(1).! ).map { CharSeq(_) }
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:429]: i'm worried that everything is a def rather than a val: the number of instanciated objects will explode.

Comment (ncharles) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:429]: see https://github.com/Normation/rudder/pull/2467

Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:429]: I'm asking on fastparse how to handle that case

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/Constraint.scala [hunk 2]
```diff
@@ -157,12 +173,15 @@ object MailVType extends FixedRegexVType {
...
-      (try {
-        Right(value.toInt)
-      } catch {
-        case _:Exception => Left(Inconsistency(s"Wrong value ${value} for field '${forField}': expecting an integer."))
-      }).map( _ => value )
+    super.getFormatedValidated(value, forField, escapeString).flatMap( escaped =>
+      VTypeConstraint.checkIsProperty(value, forField) { v =>
+        try {
+          v.toInt
+          Right(())
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/Constraint.scala:180]: why is it returning an empty Right ?

Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/Constraint.scala:180]: to force the user (me) to `map(_ => escaped)` for the return value. We don't use these value, the parsing/etc is just done to check if the provided input is of the correct type, but we need to use the escaped string char that is already computed before that method is called

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/RegexConstraint.scala [hunk 3]
```diff
@@ -52,15 +54,26 @@ case class RegexConstraint(pattern: String, errorMsg: String) {
 
   val compiled = Pattern.compile(pattern)
 
-  val variablePattern = Pattern.compile(".*?\\$\\{.*?\\}.*?")
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/RegexConstraint.scala:55]: i like that it is removed :)

**Comment by ncharles** [comment]
this looks good to me, but I'm not sure i'll be able to understand the parsing part

**Comment by fanf** [comment]
Fortunately, there is a bunch of test cases. It seems that whoever wrote the first parser had had to deal with a great deal of edge cases. Or he was just a bit paranoid. They helped a lot and caught a lot of problem when adding the `${whatever}` case

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 1 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:429]: https://gitter.im/lihaoyi/fastparse?at=5ea9e817f0377f163158e8b4
```
Since fastparse 2.x, parsers no longer have any initialization cost. They are simply methods that you call just like any other methods
```

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 4]
```diff
@@ -406,64 +405,88 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
+  def emptySQuote[_: P] : P[DefaultValue] = P( "\"\"" ).map { _ => DefaultValue(CharSeq("")::Nil) }
+  def emptyTQuote[_: P] : P[DefaultValue] = P("\"\"\"\"\"\"").map { _ => DefaultValue(CharSeq("")::Nil) }
 
-  //string must be simple or triple quoted string
-  val sqString    : Parser[DefaultValue] = id("\"")  ~> rep1(sqplainStr | interpol )<~ id("\"") ^^ { case x => DefaultValue(x) }
-  val sqplainStr  : Parser[Token]        = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties|")).)+""").r  ^^ { str => CharSeq(str) }
 
-  val tqString    : Parser[DefaultValue]   = id("\"\"\"")  ~> rep1(tqplainStr | interpol )<~ id("\"\"\"") ^^ { case x => DefaultValue(x) }
-  val tqplainStr  : Parser[Token]          = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties|"""+"\"\"\""+ """)).)+""").r  ^^ { str => CharSeq(str) }
+  def leftTag[_: P] = P( "<" ~ (!">" ~ AnyChar).rep(1).! ~ ">" )
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:484]: unused, can be removed

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 5]
```diff
@@ -406,64 +405,88 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
-
+  def emptyVar[_: P]: P[List[CharSeq]] = P(CharsWhile(_.isSpaceChar, 0).!).map(x => List(CharSeq(x)) )
+  def s[_:P] = P(CharsWhile(_.isWhitespace, 0))
   // plain string must not match our identifier, ${rudder.* and ${node.properties.*}
-  // here we defined a function to build them, with the possibility to
-  val plainString : Parser[CharSeq] = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties)).)+""").r  ^^ { CharSeq(_) }
+  // here we defined a function to build them
+  def plainStringNoDollar[_: P] : P[CharSeq] = P( (!"${" ~ AnyChar).rep(1).! ).map { CharSeq(_) }
+  def plainStringNoEndCurly[_: P] : P[CharSeq] = P( (!"}" ~ AnyChar).rep(1).! ).map { CharSeq(_) }
+  def plainStringNoStartCurly[_: P] : P[CharSeq] = P( (!"{" ~ AnyChar).rep(1).! ).map { CharSeq(_) }
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:431]: unused, can be removed

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 6]
```diff
@@ -406,64 +405,88 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
+  }
 
-  val space = """\s*""".r
+  def all[_: P] : P[List[Token]] = P( Start ~ ( (plainStringNoDollar | varOrString).rep(1) | emptyVar) ~ End).map(_.toList)
 
   //empty string is a special case that must be look appart from plain string.
   //also parse full blank string, because no need to look for more case for them.
-  val emptyVar    : Parser[List[CharSeq]] = """(?iums)(\s)*""".r ^^ { x => List(CharSeq(x)) }
-
+  def emptyVar[_: P]: P[List[CharSeq]] = P(CharsWhile(_.isSpaceChar, 0).!).map(x => List(CharSeq(x)) )
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: you take only space character here, 

You could write 

s.!.map(x => List(CharSeq(x)))

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 7]
```diff
@@ -406,64 +405,88 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
-  val sqString    : Parser[DefaultValue] = id("\"")  ~> rep1(sqplainStr | interpol )<~ id("\"") ^^ { case x => DefaultValue(x) }
-  val sqplainStr  : Parser[Token]        = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties|")).)+""").r  ^^ { str => CharSeq(str) }
 
-  val tqString    : Parser[DefaultValue]   = id("\"\"\"")  ~> rep1(tqplainStr | interpol )<~ id("\"\"\"") ^^ { case x => DefaultValue(x) }
-  val tqplainStr  : Parser[Token]          = ("""(?iums)((?!(\Q${\E\s*rudder\s*\.|\Q${\E\s*node\s*\.\s*properties|"""+"\"\"\""+ """)).)+""").r  ^^ { str => CharSeq(str) }
+  def leftTag[_: P] = P( "<" ~ (!">" ~ AnyChar).rep(1).! ~ ">" )
+
+  //string must be simple or triple quoted string
+  def sqString[_: P]    : P[DefaultValue] = P( "\"" ~ (sqplainStr | variable ).rep(1) ~ "\"").map { case x => DefaultValue(x.toList) }
+  def sqplainStr[_: P]  : P[Token]        = P( CharsWhile(c => c != '"').! ).map { str => CharSeq(str) }
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:488]: This does not managed escaped quote

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:488]: also, i thin it will ignores variables, it should be something like the begining 

(plainStringNoDollar | varOrString)

But cannot use them, because the condition of underlying strings are not the same (no quote character)

**Comment by VinceMacBuche** [comment]
Also I think you should rename s to space, put long names instead of shortname (ie: sqStr, should be singleQuoteString)

**Comment by VinceMacBuche** [comment]
I started a PR of renaming, a removal, also adding some edge cases tests

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 6 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: Also I think it is already treated by plainStringNoDollar

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 6 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: no, it's needed, I added a test case for it

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 7 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:488]: we never managed escaped quote in node properties default, we have triple quote for that. If you think of a use case node supported, I should go in an other bug report I believe

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 6 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: i removed it and added a test case and all tests are ok

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 5 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala]: corrected

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 4 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala]: corrected

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 7 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:484]: ok no need to support escaped quote, if it was not like before 

But you are ignoring variables in string 

default=" defaut ${node.properties[datacenterDefault]}" should match Default(CharSeq(" defaut "), Property(datacenterDefault)) but it is a CharSeq" defaut ${node.properties[datacenterDefault]}"

I added a test case in my PR

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 7 above)
Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:486]: That pr fixes mosr problems with inner variables in quotes, i also need to look some other edge cases (non ending variables, non Rudder variable too https://github.com/fanf/rudder/pull/10/files

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala [hunk 8]
```diff
@@ -406,64 +405,83 @@ class InterpolatedValueCompilerImpl extends RegexParsers with InterpolatedValueC
...
+  }
 
-  val space = """\s*""".r
+  def all[_: P] : P[List[Token]] = P( Start ~ ( (noVariableStart | variableStart).rep(1) | emptyVar) ~ End).map(_.toList)
 
   //empty string is a special case that must be look appart from plain string.
   //also parse full blank string, because no need to look for more case for them.
-  val emptyVar    : Parser[List[CharSeq]] = """(?iums)(\s)*""".r ^^ { x => List(CharSeq(x)) }
-
+  def emptyVar[_: P]: P[List[CharSeq]] = P(CharsWhile(_.isSpaceChar, 0).!).map(x => List(CharSeq("")) )
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: Ha ok it's just empty ?? ok I thought is was whitspace ?

**Comment by fanf** [comment]
PR updated with a new commit

File: webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala (hunk 6 above)
Comment (fanf) [reply webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/InterpolatedValueCompiler.scala:425]: the problem is with empty string `""`

**Comment by VinceMacBuche** [comment]
OK, squash merging this PR
